rust/otap-dataflow/crates/otap/src/otap_batch_processor.rs

```diff
@@ -0,0 +1,922 @@
+                }
+            }
+        }
+        if !self.passthrough.is_empty() {
+            let mut remaining = Vec::with_capacity(self.passthrough.len());
+            for pdata in self.passthrough.drain(..) {
+                if pdata.signal_type() == SignalType::Traces {
+                    effect.send_message(pdata).await?;
+                } else {
+                    remaining.push(pdata);
+                }
+            }
+            self.passthrough = remaining;
+        }
+        self.rows_traces = 0;
+        Ok(())
+    }
+}
+
+/// Factory function to create an OTAP batch processor
+pub fn create_otap_batch_processor(
+    _pipeline_ctx: otap_df_engine::context::PipelineContext,
+    node: NodeId,
+    config: &Value,
+    processor_config: &ProcessorConfig,
+) -> Result<ProcessorWrapper<OtapPdata>, ConfigError> {
+    OtapBatchProcessor::from_config(node, config, processor_config)
+}
+
+#[async_trait(?Send)]
+impl local::Processor<OtapPdata> for OtapBatchProcessor {
+    async fn process(
+        &mut self,
+        msg: Message<OtapPdata>,
+        effect: &mut local::EffectHandler<OtapPdata>,
+    ) -> Result<(), EngineError> {
+        match msg {
+            Message::Control(ctrl) => {
+                match ctrl {
+                    otap_df_engine::control::NodeControlMsg::TimerTick { .. } => {
+                        // Follow OTLP semantics: only flush on timer if a size threshold has been reached
+                        if self.size_triggers_emission() {
+                            self.flush_current(effect).await
+                        } else {
+                            Ok(())
+                        }
+                    }
+                    otap_df_engine::control::NodeControlMsg::Config { .. } => Ok(()),
+                    otap_df_engine::control::NodeControlMsg::Shutdown { .. } => {
+                        // Flush and shutdown
+                        self.flush_current(effect).await?;
+                        effect.info(LOG_MSG_SHUTTING_DOWN).await;
+                        Ok(())
+                    }
+                    otap_df_engine::control::NodeControlMsg::Ack { .. }
+                    | otap_df_engine::control::NodeControlMsg::Nack { .. }
+                    | otap_df_engine::control::NodeControlMsg::CollectTelemetry { .. } => Ok(()),
+                }
+            }
+            Message::PData(data) => {
+                let max = self.config.send_batch_max_size;
+                match OtapArrowRecords::try_from(data.clone()) {
+                    Ok(rec) => {
+                        let rows = rec.batch_length();
+                        // Per-signal policy: drop zero-row; pre-flush if exceeding max; append rows; flush on max or send_batch_size.
+                        match &rec {
+                            OtapArrowRecords::Logs(_) => {
+                                if rows == 0 {
+                                    effect.info(LOG_MSG_DROP_NON_OTAP).await;
```

Comment:
This `LOG_MSG_DROP_NON_OTAP ` message is:
```
"OTAP batch processor: dropping non-OTAP message (conversion failed)"
```
But that's not actually what happened here, right? It's not that we failed to convert to OTAP, it's that the batch was  0 length.

Reply:
thanks for correcting the error message

Reply:
-  Zero-row batches now log LOG_MSG_DROP_EMPTY ("OTAP batch processor: dropping empty OTAP record (0 rows)") and are dropped.
- Conversion failures log LOG_MSG_DROP_CONVERSION_FAILED ("OTAP batch processor: dropping message: OTAP conversion failed") and are dropped.
- The old LOG_MSG_DROP_NON_OTAP is gone now